File: src/libs/rust/base/src/arch/linux/ioctl.rs
use libc;

use crate::cfg;
use crate::kif;
use crate::mem::VirtAddr;
use crate::tcu::ActId;

use super::tcu_fd;

// this is defined in linux/drivers/tcu/tcu.cc
const IOCTL_WAIT_ACT: u64 = 0x80087101;
const IOCTL_RGSTR_ACT: u64 = 0x40087102;
const IOCTL_TLB_INSERT: u64 = 0x40087103;
const IOCTL_UNREG_ACT: u64 = 0x40087104;
const IOCTL_NOOP: u64 = 0x00007105;

fn ioctl(magic_number: u64) {
    unsafe {
        let res = libc::ioctl(tcu_fd(), magic_number);
        if res != 0 {
            libc::perror(0 as *const u8);
            panic!("ioctl call {} failed with error {}", magic_number, res);
        }
    }
}

fn ioctl_read<T: Default>(magic_number: u64) -> T {
    let mut arg: T = T::default();
    unsafe {
        let res = libc::ioctl(tcu_fd(), magic_number, &mut arg as *mut _);
        if res != 0 {
            libc::perror(0 as *const u8);
            panic!("ioctl call {} failed with error {}", magic_number, res);
        }
    }
    arg
}

fn ioctl_plain(magic_number: u64, arg: usize) {
    unsafe {
        let res = libc::ioctl(tcu_fd(), magic_number, arg);
        if res != 0 {
            libc::perror(0 as *const u8);
            panic!("ioctl call {} failed with error {}", magic_number, res);
        }
    }
}

pub fn wait_act() -> ActId {
    ioctl_read(IOCTL_WAIT_ACT)
}

pub fn register_act(id: ActId) {
    ioctl_plain(IOCTL_RGSTR_ACT, id as usize);
}

pub fn tlb_insert_addr(virt: VirtAddr, perm: u8) {
    // touch the memory first to cause a page fault, because the TCU-TLB miss handler in the Linux
    // kernel cannot deal with the request if the page isn't mapped.
    let virt_ptr = virt.as_mut_ptr::<u8>();
    if (perm & kif::Perm::W.bits() as u8) != 0 {
        unsafe { virt_ptr.write_volatile(0) }
    }
    else {
        let _val = unsafe { virt_ptr.read_volatile() };
    }

    let arg = virt.as_local() & !cfg::PAGE_MASK | perm as usize;
    ioctl_plain(IOCTL_TLB_INSERT, arg);
}

pub fn unregister_act(id: ActId) {
    ioctl_plain(IOCTL_UNREG_ACT, id as usize);
}

pub fn noop() {
    ioctl(IOCTL_NOOP);
}

File: src/libs/base/arch/linux/IOCtl.cc
/*
 * Copyright (C) 2023 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/KIF.h>
#include <base/Panic.h>
#include <base/arch/linux/IOCtl.h>

#include <sys/ioctl.h>

namespace m3lx {

// this is defined in linux/drivers/tcu/tcu.cc
static const ulong IOCTL_WAIT_ACT = 0x80087101;
static const ulong IOCTL_RGSTR_ACT = 0x40087102;
static const ulong IOCTL_TLB_INSERT = 0x40087103;
static const ulong IOCTL_UNREG_ACT = 0x40087104;
static const ulong IOCTL_NOOP = 0x00007105;

void tlb_insert_addr(uintptr_t addr, uint perm) {
    using namespace m3;

    // touch the memory first to cause a page fault, because the TCU-TLB miss handler in the Linux
    // kernel cannot deal with the request if the page isn't mapped.
    UNUSED uint8_t dummy;
    volatile uint8_t *virt_ptr = reinterpret_cast<uint8_t *>(addr);
    if(perm & KIF::Perm::W)
        *virt_ptr = 0;
    else
        dummy = *virt_ptr;

    size_t arg = (addr & ~PAGE_MASK) | perm;
    int res = ::ioctl(tcu_fd(), IOCTL_TLB_INSERT, arg);
    if(res != 0)
        panic("ioctl call TLB_INSERT failed\n"_cf);
}

}
